src/zcash/Address.hpp

```diff
@@ -97,6 +99,116 @@ typedef boost::variant<InvalidEncoding, SproutPaymentAddress> PaymentAddress;
 typedef boost::variant<InvalidEncoding, SproutViewingKey> ViewingKey;
 typedef boost::variant<InvalidEncoding, SproutSpendingKey> SpendingKey;
 
+//! Sapling functions. 
+class SaplingPaymentAddress {
+public:
+    diversifier_t d;
+    uint256 pk_d;
+
+    SaplingPaymentAddress() : d(), pk_d() { }
+    SaplingPaymentAddress(diversifier_t d, uint256 pk_d) : d(d), pk_d(pk_d) { }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(d);
+        READWRITE(pk_d);
+    }
+
+    friend inline bool operator==(const SaplingPaymentAddress& a, const SaplingPaymentAddress& b) {
+        return a.d == b.d && a.pk_d == b.pk_d;
+    }
+    friend inline bool operator<(const SaplingPaymentAddress& a, const SaplingPaymentAddress& b) {
+        return (a.d < b.d ||
+                (a.d == b.d && a.pk_d < b.pk_d));
+    }
+};
+
+class SaplingInViewingKey : public uint256 {
+public:
+    SaplingInViewingKey() : uint256() { }
+    SaplingInViewingKey(uint256 ivk) : uint256(ivk) { }
+    
+    // Can pass in diversifier for Sapling addr
+    SaplingPaymentAddress address(diversifier_t d) const;
```

Comment:
ditto. This should return an optional or something. We only ever call it with a successful diversifier from default_diversifier, but that won't be the case forever.